backend/src/org/commcare/suite/model/OfflineUserRestore.java

```diff
@@ -0,0 +1,135 @@
+package org.commcare.suite.model;
+
+import org.javarosa.core.io.StreamsUtil;
+import org.javarosa.core.reference.InvalidReferenceException;
+import org.javarosa.core.reference.Reference;
+import org.javarosa.core.reference.ReferenceManager;
+import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.util.PropertyUtils;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+import org.javarosa.xml.ElementParser;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+
+/**
+ * User restore xml file sometimes present in apps.
+ * Used for offline (demo user) logins.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ * @author Aliza Stone (astone@dimagi.com)
+ */
+public class OfflineUserRestore implements Persistable {
+    public static final String STORAGE_KEY = "OfflineUserRestore";
+    private int recordId = -1;
+    private String restore;
+    private String reference;
+    private String username;
+    private String password;
+
+    public OfflineUserRestore() {
+    }
+
+    public OfflineUserRestore(String reference) {
+        this.reference = reference;
+        this.username = parseUsernameFromRestoreFile(reference);
+        this.password = PropertyUtils.genUUID();
+    }
+
+    public static OfflineUserRestore buildInMemoryUserRestore(InputStream restoreStream) throws IOException {
+        OfflineUserRestore offlineUserRestore = new OfflineUserRestore();
+        offlineUserRestore.restore = new String(StreamsUtil.inputStreamToByteArray(restoreStream));
+        return offlineUserRestore;
```

Comment:
I sorta wonder if for parity it would make sense to set the username/password in this builder, similar to how it is done in `OfflineUserRestore(String reference)`